### PR TITLE
fix: possible fix for isTTY and GitBash

### LINKED ISSUE #19: Possible typo in the source code
Heya. As ususal, i always review the whole code of libraries or whatever.

So found that the `this.rl` here https://github.com/terkelg/prompts/blob/230c1a674965a065f931012be74af3ca523f1fb0/lib/elements/prompt.js#L18-L19

probably is meant to be the `rl` const, because the `this.rl` is not defined in any other place in the code base, even in that file.

### LINKED ISSUE #35: Prompts doesn't work in global packages with Git Bash or Windows CMD
**OS:** Windows 7 Ultimate 64-bit SP1
**Node**: v9.5.0
**Prompts**: v0.1.4

Whenever I press 'enter' on a `prompts.select` or `prompts.text`, I get:
```shell
TypeError: this.in.setRawMode is not a function
    at TextPrompt.close (C:\projects\_PLAYGROUND\PromptsDontWorkGlobal\node_modules\prompts\lib\elements\prompt.js:39:15)
    at TextPrompt.submit (C:\projects\_PLAYGROUND\PromptsDontWorkGlobal\node_modules\prompts\lib\elements\text.js:50:10)
    at Socket.keypress (C:\projects\_PLAYGROUND\PromptsDontWorkGlobal\node_modules\prompts\lib\elements\prompt.js:30:16)
    at Socket.emit (events.js:160:13)
    at emitKeys (internal/readline.js:420:14)
    at emitKeys.next (<anonymous>)
    at Socket.onData (readline.js:1021:36)
    at Socket.emit (events.js:165:20)
    at addChunk (_stream_readable.js:269:12)
    at readableAddChunk (_stream_readable.js:256:11)
```

**CODE:**
```javascript
#!/usr/bin/env node

const prompts = require('prompts');

const presets = {
    'App': 'OTHERDATA',
    '(Express|EJS)': 'OTHERDATA',
    'Webpack-Dev-Server + (Webpack|ES6-BABEL|SCSS)': 'OTHERDATA',
    'React + Webpack-Dev-Server + (Webpack|ES6-BABEL)': 'OTHERDATA',
    'React + Redux + Webpack-Dev-Server + (Webpack|ES6-BABEL)': 'OTHERDATA'
};

(async () => {
    const { answer } = await prompts({
        type: 'text',
        name: 'answer',
        message: 'Hello FöRVaiS'
    });

    console.log(answer);

    const { data: [, data] } = await prompts({
        type: 'select',
        name: 'data',
        message: 'Select a preset from this list',
        choices: Object.entries(presets).map(preset => ({ title: preset[0], value: preset })),
        initial: 0
    });

    console.log(data);
})();
```

### PR DESCRIPTION
**Pull request by olstenlarck**
fixes #35, closes #19

### CODE REVIEW

**Comment by FoRVaiS** [comment]
7c042e3: this.clear doesn't exist
```
$ /c/projects/_PLAYGROUND/PromptsPR#36/index.js
? Hello FöRVaiS » Hello!
√ Hello FöRVaiS ... Hello!
readline.js:1031
            throw err;
            ^

TypeError: this.clear is not a function
    at TextPrompt.close (C:\projects\_PLAYGROUND\PromptsPR#36\node_modules\prompts\lib\elements\prompt.js:45:17)
    at TextPrompt.submit (C:\projects\_PLAYGROUND\PromptsPR#36\node_modules\prompts\lib\elements\text.js:50:10)
    at Socket.keypress (C:\projects\_PLAYGROUND\PromptsPR#36\node_modules\prompts\lib\elements\prompt.js:30:16)
    at Socket.emit (events.js:160:13)
    at emitKeys (internal/readline.js:420:14)
    at emitKeys.next (<anonymous>)
    at Socket.onData (readline.js:1021:36)
    at Socket.emit (events.js:165:20)
    at addChunk (_stream_readable.js:269:12)
    at readableAddChunk (_stream_readable.js:256:11)
```

**Comment by FoRVaiS** [comment]
Cause of Bug
-------------
In a TTY environment, such as VS Code's integrated terminal, Prompts works as intended.

In non-TTY environments like Git Bash:
If you (implicitly)? set the interpreter, `#!/usr/bin/env node` then run the file `./promptsTest.js` Prompts starts to misbehave.

The cause of the bug comes from implicitly calling Node with a shebang in a non-TTY environment.
Thus running a file with an absolute path `/c/.../promptsTest.js` or 
with a bin command with this configuration: 
```json
"bin": {
    "prompts-test": "promptsTest.js"
}
```
then running the command
```shell
prompts-test
```

will both result in a broken Prompts application.

Workaround
-------------
Explicitly run a file with Node: `node promptsTest.js`.
I cannot find a workaround for bin commands.

**Comment by olstenlarck** [comment]
> nothing there is needed.

@lukeed what's not needed? :laughing: `*.log` files may appear always. Coverage is up coming any way. Plus the NPM's lockfile.

> Looks like you forgot to import from util file.

It was meant to be `this.clear`, but obviously we should try with `clear` from util.

> I cannot find a workaround for bin commands.

@FoRVaiS, what about

```json
"bin": {
  "prompts-test": "node promptsTest.js"
}
```

never tried it.

**Comment by FoRVaiS** [comment]
```json
"bin": {
  "prompts-test": "node promptsTest.js"
}
```
left me an error saying that it couldn't find promptsTest.js. I fail to understand why that happens, since calling the file directly `promptsTest.js` would work perfectly (assuming we fix the non-TTY bug).

**Comment by olstenlarck** [comment]
Try with absolute path or `./`.

I had problems few times when just defining 

```
"bin": {
  "foobar": "foobar.js"
}
```

So since then i always prepend `./` when doing CLIs.

**Comment by FoRVaiS** [comment]
```json
"bin": {
  "prompts-test": "node ./promptsTest.js"
}
```
outputs this npm err. I don't get the chance to even run the test
```shell
$ npm publish; npm install -g local-promptstest
+ local-promptstest@0.0.5
npm ERR! path C:\Program Files\nodejs\node_modules\local-promptstest\node .\index.js
npm ERR! code ENOENT
npm ERR! errno -4058
npm ERR! syscall chmod
npm ERR! enoent ENOENT: no such file or directory, chmod 'C:\Program Files\nodejs\node_modules\local-promptstest\node .\index.js'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\${user}\AppData\Roaming\npm-cache\_logs\2018-03-03T02_22_02_884Z-debug.log
```

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,10 @@
 node_modules
 .DS_Store
 *.lock
+*.log
```
Comment (terkelg) [inline .gitignore]: Please remove .gitignore from this commit. The ignored files are not related to this project. Include them in your own global gitignore if needed.

File: lib/elements/prompt.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ class Prompt extends EventEmitter {
     this.out = process.stdout;
 
     const rl = readline.createInterface(this.in);
-    readline.emitKeypressEvents(this.in, this.rl);
+    readline.emitKeypressEvents(this.in, rl);
```
Comment (terkelg) [inline lib/elements/prompt.js]: I still have to look into this. The second param to `emitKeypressEvents` is optional - do we even need it?

Also have to dig deeper into `createInterface` options like `crlfDelay` and `terminal`
https://nodejs.org/api/readline.html#readline_readline_createinterface_options

**Comment by terkelg** [comment]
How can I test this on a mac?

**Comment by FoRVaiS** [comment]
This looks like a rendering issue. 

In this demo, pressing the down arrow twice lands me at the correct option but since it never rerenders, we see it stuck in the initial position. I added an output to report on the cursor's location but it never renders when `prompt.select` is up. They end up rendering after I end `prompt.select`.

demo3

Ignore the `this.cursor + 1`, forgot to remove it :P

**Comment by FoRVaiS** [comment]
Based on my 'bad rendering' theory, I've traced a potential problem to `sisteransi.erase`.

Does anyone happen to know if using `#!/usr/bin/env node` at the top of a file is different in anyway compared to doing `node promptsTest.js`?

If there is, maybe these codes don't work when using a shebang?
```javascript
const erase = {
  screen: `${ESC}2J`,
  up: `${ESC}1J`,
  down: `${ESC}J`,
  line: `${ESC}2K`,
  lineEnd: `${ESC}K`,
  lineStart: `${ESC}1K`,
  lines(count) {
    let clear = '';
    for (let i = 0; i < count; i++)
      clear += this.line + (i < count - 1 ? cursor.up() : '');
    if (count)
      clear += cursor.left;
    return clear;
  }
}
```

**Comment by olstenlarck** [comment]
> maybe these codes don't work when using a shebang?

@FoRVaiS i think that's the diff, yea.

**Comment by olstenlarck** [comment]
It's really frustrating to play on guesses...

**Comment by FoRVaiS** [comment]
Email notifications are broken :x sorry

index

I can not even begin to explain what is going on here.

**Comment by olstenlarck** [comment]
Thanks! So `text` and `select` are okey now?

**Comment by olstenlarck** [comment]
Thanks! Cool, starting to see fix :D

**Comment by FoRVaiS** [comment]
Nope, text and select are not okay yet.
- `prompts.select` remains unscrollable (my theory is failure to re-render)
- Both `prompts.select` and `prompts.text` fail to clear their previous lines

File: lib/elements/prompt.js [hunk 3]
```diff
@@ -39,8 +39,8 @@ class Prompt extends EventEmitter {
       if (this.in.isTTY) {
         this.in.setRawMode(false);
       } else {
-        // we can't use `utils.clear()`, because it expect `prompt` string
-        this.out.write(erase.line + cursor.prevLine());
+        this.clear = clear(this.prompt)
```
Comment (olstenlarck) [inline lib/elements/prompt.js]: `this.prompt` should be available, because `this.close` is called after the `this.render` inside the `submit()`...

**Comment by FoRVaiS** [comment]
Funny thing, it fixed the 'previous line' for `prompt.select` (still remains unscrollable) but didn't fix 'previous line' for `prompt.text`

screenshot_2

**Comment by FoRVaiS** [comment]
4fdaf6e No change.

**Comment by olstenlarck** [comment]
Does anyone know is AppVeyor using GitBash or such? Maybe adding AppVeyor would help the things a bit.

But in any way, such problems are hard cuz is user interaction..

**Comment by FoRVaiS** [comment]
So using both `node sisteransi.js` & `./sisteransi.js` I was able to get a consistent result. Both types of execution for sure do not affect the way unicode behave. ( Probably why it's called UNI-code :P )
```javascript
#!/usr/bin/env node

const { erase } = require('sisteransi');

const out = process.stdout;

console.log('Test 1');
console.log('Test 2');
console.log('Test 3');
console.log('Test 4');
console.log('Test 5');
console.log('Test 6');
out.write(erase.lines(3 + 1));
```
Expected Output:
--------
Test 1
Test 2
Test 3

Output:
--------
Test 1
Test 2
Test 3

**Comment by FoRVaiS** [comment]
I seriously don't get it... it seems as if render() is being queued or delayed? 

TTY Environment/Terminal
---------------------------
demo5


Non-TTY Environment/Terminal
--------------------------------
demo4

**Comment by FoRVaiS** [comment]
To demonstrate the 'queuing' behavior, I've modified: 

https://github.com/terkelg/prompts/blob/5b54227e506a5e54ac9f32a56b62167f4d1d07e9/lib/elements/text.js#L96
to:
`this.out.write('\n' + prompt); `

Result:
------
demo6

@terkelg @olstenlarck does this bring us any closer to finding a solution?

**Comment by FoRVaiS** [comment]
@terkelg @olstenlarck So I'm 110% sure I've found the source of the problem. The problem is caused by readline not reading keypress events in non-TTY instances.

Main thread to read:
https://github.com/nodejs/node/issues/2996

The queuing issue is from the keypress event waiting for a user to input the 'ENTER' key first before starting to read inputs.

This person gets the same queuing problem:
https://stackoverflow.com/questions/5006821/nodejs-how-to-read-keystrokes-from-stdin

**Comment by terkelg** [comment]
@FoRVaiS thank you for doing so much research on this one. Appreciate it.

**Comment by FoRVaiS** [comment]
@terkelg No worries, if a workaround/fix can be found, then it benefits us both. Otherwise, at least you would know what the exact problem is, when someone else has the same problem. :P

I'm unable to think of a workaround. The only solution that I can come up with, that doesn't require a redesign of the input system, would be to simulate an 'enter' keypress but all the packages that do keyboard simulation require Java's Robot package. 

If a workaround is proposed, I'd be glad to try it out.

**Comment by olstenlarck** [comment]
Great! :tada: So much thank you for that!

In that case, I think we definitely should close the PR. It's not specific to this prompt module, but to Git Bash, so not make sense to touch on our side.

Adding notice on the README and pointing to this PR or to #35 issue.
